src/hotspot/share/gc/g1/g1ConcurrentRefine.cpp

```diff
@@ -117,37 +110,36 @@ jint G1ConcurrentRefineThreadControl::initialize(G1ConcurrentRefine* cr, uint ma
 
 #ifdef ASSERT
 void G1ConcurrentRefineThreadControl::assert_current_thread_is_primary_refinement_thread() const {
-  assert(_threads != nullptr, "No threads");
-  assert(Thread::current() == _threads[0], "Not primary thread");
+  assert(Thread::current() == _threads.at(0), "Not primary thread");
 }
 #endif // ASSERT
 
 bool G1ConcurrentRefineThreadControl::activate(uint worker_id) {
-  assert(worker_id < _max_num_threads, "precondition");
-  G1ConcurrentRefineThread* thread_to_activate = _threads[worker_id];
+  assert(worker_id < max_num_threads(), "precondition");
+  G1ConcurrentRefineThread* thread_to_activate = _threads.at(worker_id);
   if (thread_to_activate == nullptr) {
     thread_to_activate = create_refinement_thread(worker_id, false);
     if (thread_to_activate == nullptr) {
       return false;
     }
-    _threads[worker_id] = thread_to_activate;
+    _threads.at(worker_id) = thread_to_activate;
```

Comment:
With size-based tracking and the suggested helper function, replace this whole function with
```
  if (ensure_threads_created(worker_id, false)) {
    _threads.at(worker_id)->activate();
    return true;
  } else {
    return false;
  }
```